### ISSUE TITLE
todo_comment_linter could get an argument for valid TODO patterns

### ISSUE THREAD

**Issue by MichaelChirico** [r-lib/lintr#2047]
This can allow `TODO` _only if_ they fit certain patterns.

This could be used to enforce a rule like "All TODO must be tied to a bug, not a person".

```r
## BAD
# TODO(michaelchirico): fix this
# TODO: add testing

## GOOD
# TODO(#nnn): fix this
# TODO(#mmm): add testing
```

Maybe an argument like `allow_regex` or `allow_style_regex`, in this case `allow_regex = "TODO[(][#][0-9]+[)]:"` could work.

**Comment by AshesITR** [comment]
Maybe just expose the regex as an optional argument and call it `regex=`?

`todo_comment_linter(regex = "# TODO\\(#\\d+\\):")` seems fine to me.

**Comment by MichaelChirico** [comment]
The question then is how to interact this with the existing `todo` argument:

https://github.com/r-lib/lintr/blob/191e3b7638218920f90688dc6b96fe69c05f4a8a/R/comment_linters.R#L164

Should we be looking for matches of `paste0(todo, regex)`? Then the call would be `todo_comment_linter(regex = "\\(#\\d+\\):")`.

**Comment by AshesITR** [comment]
I would let `regex` take precedence with a warning if both arguments are present.

**Comment by MichaelChirico** [comment]
After #2433, I'm thinking the approach here will be:

```r
function(todo = c("todo", "fixme"), except_regex = NULL)
```

Where `except_regex` is used to filter out lints matched by `todo`. It also makes sense verbally: Don't use `todo` _unless_ it matches one of `except_regex`.

I think we should support case-sensitive matching too -- I would want `TODO(#1234)` to be OK but not `todo(#1234)`. Probably letting `todo=` be case-insenstive but `except_regex=` case-sensitive makes sense, otherwise we could add a third argument.

**Comment by AshesITR** [comment]
I don't think we need a third argument. You can simply turn on ignore-case by prepending `except_regex` with `"(?i)"` if necessary.
